Properly working (dynamic) gatsby frontend for Public People

fardarter opened the pull request:
…utility to handle the quesry string, test for utility, global css stylesheet to contain cachable styles, hopefully for refactoring, a TODO.md, and IDEAS.md, and some additions to the README.md.

fardarter committed:
Added: Individual person page, reducers to handle the request, extra utility to handle the quesry string, test for utility, global css stylesheet to contain cachable styles, hopefully for refactoring, a TODO.md, and IDEAS.md, and some additions to the README.md.

schalkventer commented on README.md:
```diff
@@ -56,26 +69,28 @@ The following Redux bindings/plugins are used:
+2.  Make sure you have [NodeJS](https://nodejs.org/en/) installed.
+3.  Run `npm install` in the root folder of the repository.
+4.  Run `npm start` to spin up the development server.\*
+5.  The 'npm start`script is configured to run on`localhost:3000` as the media/news search dev endpoint responds to this origin.
```
Seems you started with a quote here but ended with a back-tick.

schalkventer commented on README.md:
```diff
@@ -56,26 +69,28 @@ The following Redux bindings/plugins are used:
+3.  Run `npm install` in the root folder of the repository.
+4.  Run `npm start` to spin up the development server.\*
+5.  The 'npm start`script is configured to run on`localhost:3000` as the media/news search dev endpoint responds to this origin.
+6.  Open `localhost:3000` in your browser.
```
Isn't it easier to just ask @jbothma to change the cross-origin to the existing port?

jbothma replied:
is the npm start default 3000? could we leave it at 3000 then? the backend uses 8000 by default and defaults are best unless there's a reason for changing - front and backend also happen to not conflict if someone wants to work on both side-by-side in the future.

schalkventer commented on gatsby-browser.js:
```diff
@@ -1,21 +1,29 @@
 
 exports.onClientEntry = () => {
-  const phrase = extractQueryString('phrase', window.location.search);
+  // This function fires onloading the page.
```
Don't think this comment is necessarily, since it's implied by the Gatsby architecture/docs itself.

schalkventer commented on gatsby-browser.js:
```diff
@@ -27,7 +35,7 @@ exports.replaceRouterComponent = ({ history }) => {
 
   ConnectedRouterWrapper.propTypes = {
-    children: PropTypes.node.isRequired,
+    children: PropTypes.node.isRequired
```
It's generally a good idea to have trailing commas, even if they are not required: https://eslint.org/docs/rules/comma-dangle.html.

fardarter replied:
I'm going to prefer no comma as it breaks my JSON, but I have implemented this as a rule in linting.

schalkventer commented on gatsby-config.js:
```diff
@@ -1,17 +1,10 @@
 const loadFont = {
-  resolve: 'gatsby-plugin-google-fonts',
+  resolve: "gatsby-plugin-google-fonts",
```
Single or double quotes are most a stylistic preference, if you prefer double quotes just make sure that they are used consistently in the JS everywhere.

fardarter replied:
I'm going to prefer double quotes for JSON reasons and I've implemented a linting rule.

schalkventer commented on src/components/Card/index.jsx:
```diff
@@ -12,44 +11,39 @@ export default function Card(props) {
-    (link ? styles.isLink : null),
-    utils,
-  ].join(' ');
+    highlighted ? styles.isHighlighted : null,
```
Good catch! :)

schalkventer commented on src/components/Header/index.jsx:
```diff
@@ -1,20 +1,23 @@
+  mapStateToProps,
+  mapDispatchToProps
+)(Markup);
 export default Header;
```
Could probably use a newline for between these for readability.

schalkventer commented on src/components/Persons/partials/Markup.jsx:
```diff
@@ -0,0 +1,108 @@
+  const { loading, error, results, person } = props;
+  console.log("persons props", props);
+
+  if (error === "no-resuls") {
```
Like that you're using conditional returns instead, of one return with complicated logic. 👍

schalkventer replied:
Also shouldn't it be `no-results` and not `no-resuls`? :)

schalkventer commented on src/pages/index.jsx:
```diff
@@ -39,14 +39,12 @@ export default function IndexPage() {
               <Card title="Hello">
                 <p>afdgsfadsdhsgfdh</p>
-                <p>adfgafdgh</p>
+                <p>adfasdasdgafdgh</p>
```
:trollface:

schalkventer commented on src/utilities/js/extractFirstLastWords.js:
```diff
@@ -0,0 +1,14 @@
+export default function extractFirstLastWords(fullString) {
```
Great work on this, and the Jest test for this!

schalkventer commented:
Super impressed with what I've seen thus far! 👏 

Just one suggestion in terms of coding style, I'd suggest that keep with the existing Airbnb style guide (this is why you got linting errors), or replace it with another existing coding style. 

Think this comes down to the same discussion we had regarding existing design-systems/Material design earlier. It's easier for people to collaborate when there is an explicit shared way of doing things instead of having personal preference be the guiding factor.

If the above seems a bit heavy handed, perhaps look into integrating something like Prettier. 

Having a shared coding style across a project really speeds up code reviews and diffing in pull requests.

schalkventer commented:
Give me a shout if there is anything else you want me to look at.

fardarter committed:
Added FOR MERGE:
Individual person page, reducers to handle the request, extra utility to handle the query string, test for utility, global css stylesheet to contain cachable styles, hopefully for refactoring, a TODO.md, and IDEAS.md, and some additions to the README.md.
Have also added config for stylelint, eslint and jest, and have passed tests in the above.
Have removed backstop.js as we were having problems. Could do with putting it back sometime.
Have resolved Schalk's PR comments.

jbothma commented:
cool!

bit of a bug when selecting a person. to reproduce:

1. search cyril
2. click cyril johannes mack
3. url changes to http://localhost:3000/person?person=Cyril%20Mack but content is apparently for "Njabulo Cyril Ngubane"
4. click back and select a different person, still the same content

jbothma commented:
I'd recommend using the unique ID to select the person in the query string rather than the name. Name is something we know isn't unique.

I like pretty URLs, so I use a fake ID which is {{ id }}-{{ full-name-slug }} - my router then parses out the numeric ID and loads the data for that.

That might be one way to do it here. But at the very least we should show the right person.

jbothma commented on README.md:
```diff
@@ -56,30 +91,36 @@ The following Redux bindings/plugins are used:
+#### Contributors (to date)
+
+- Saul
+- JD Botha
```
I kinda feel github gives this to us for free, but I like the idea of highlighting contributors - some day I want to look at a nice way to thank people very publicly, perhaps some kind of badge or whatever people appreciate

jbothma commented on TODO.md:
```diff
@@ -0,0 +1,3 @@
+# Bugs
```
I'd suggest putting this somewhere that's actually part of a workflow, like the project trello. Which reminds me, I should add you to the project trello! done

fardarter committed:
Amended utilitiy classes, added one of my own, added graphql for meta, refactored Helmet component into header, added layour utility class "dist"

fardarter committed:
BounceWrape replaced. Card elements created

fardarter committed:
Componentised Card

fardarter committed:
Some proptypes addded. Some flex issues fixed

fardarter committed:
Resolved fetch for ie11

fardarter committed:
Debugs proptypes console error

fardarter committed:
More proptype errors fixed

fardarter committed:
Corrected error state handling.

fardarter committed:
Added nav to header

fardarter committed:
Promise cancellation working and peopleFetchWrapper both intergrating the rigght data and is cancellable.

fardarter committed:
Code for personFetch wrapper mostly done. Need to still add membership sort and pagination effects (limit, offset)

fardarter committed:
Results returned to reducer

fardarter committed:
person fetch updated and commented.

fardarter committed:
Sorted out bugs with personFetch

fardarter committed:
fetches in reducers now only fetch once

fardarter committed:
pagination in general working but bugs need resolving for: person page, for when there are few pages, and for page reload.

fardarter committed:
Some debugging. Going to move to another branch for a bit.

fardarter committed:
Apparently working.

fardarter committed:
SSR tested

fardarter committed:
New conflict on header fixed. Redux observable operational

fardarter committed:
Data fetching working in concept

fardarter committed:
Redux rearchitected and largely rewired to the UI

fardarter committed:
back button calling fetches but not updating state.

fardarter committed:
Update. Page updates now correct

fardarter committed:
Most tests passed

fardarter committed:
All linting and tests passed

fardarter committed:
Merge pull request #1 from fardarter/redux-observable

Redux observable

fardarter committed:
Various unit tests and siftedMembership function returning object with various selectors

fardarter committed:
Removed empty tests

jbothma commented on IDEAS.md:
```diff
@@ -0,0 +1,15 @@
+- Include person information (eg, memberships, current job, etc).
+- Include PDF Summary [possible tech](http://react-pdf.diegomura.com/repl)
+
+### Search Filters
```
I've added most of these to the backlog so let's remove this file.

jbothma commented on IDEAS.md:
```diff
@@ -0,0 +1,15 @@
+## Persons Page
+
+- Include person information (eg, memberships, current job, etc).
+- Include PDF Summary [possible tech](http://react-pdf.diegomura.com/repl)
```
what's the use case? why would who want this?

jbothma commented on IDEAS.md:
```diff
@@ -0,0 +1,15 @@
+- Publication
+- Date
+- Keyword
+- Locality
```
why is this important? Does locality mean of the story or publication?

jbothma commented on README.md:
```diff
@@ -1,18 +1,43 @@
+Welcome to the Public People frontend repo. We like it here and hope you will too.
 
-## Stack
+#### The Quest
```
I like the fun headings but I think it's more important to be instantly understandable, so let's be boring and call it something traditional and expected like Objective

jbothma commented on README.md:
```diff
@@ -1,18 +1,43 @@
+
+We don't mind helping – but please have a bit of a read of this document or play around a little with a local build. We assume you're a busy person with an important life (such magnificence – much grandeur) – do the same for us. One thing we might ask in exchange for our help is an update to the docs so that the next person has an easier time coming on board.
+
+*Saul (front-end gadfly)*
```
how about I add you to the mailing list and we put the project contact details here?

jbothma commented on README.md:
```diff
@@ -1,18 +1,43 @@
+
+---
+
+## Da Stack
```
Again, fun, but let's make it something like "Tech Stack" or "Technologies used"

jbothma commented on README.md:
```diff
@@ -21,33 +46,45 @@ The following Redux bindings/plugins are used:
 #### JavaScript
-- JavaScript linting done in accordance with the following ESLint configs:
+
+- Fam, use [Prettier](https://github.com/prettier/prettier) – arguing about code formatting is about the least interesting thing a person can do with their short, short life. As it can be implemented with your text editor/IDE or as a script, implementation is left up to you. Godspeed.
```
Glad you got me fam, but yeah, I'd prefer we just be boring and readable by more english speakers than the current meme followers :P

jbothma commented on src/components/Card/components/basic/Body/index.jsx:
```diff
@@ -0,0 +1,23 @@
+import styles from "./styles.module.scss";
+
+export default function BasicBody(props) {
+  // Destructuring assignment: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment
```
I don't think it makes sense to document the syntax of the language you're programming in the code. Let's state in the readme that we're using ES6 and not document syntax in the code.

fardarter committed:
Route shifts and changes bug free

fardarter committed:
Most things in place. Really need to just handle the basic pages, some colour/hover transitions and the membership list on the person page.

fardarter committed:
Minor fix.

fardarter committed:
Some styling for reading pages.

fardarter committed:
Merge branch 'master' of https://github.com/fardarter/public-people-frontend

fardarter committed:
Tidying up some error state styling.

fardarter committed:
Touch ups to person list and header

jbothma commented:
- News date is really hard to read.
- Headlines aren't aligned so they're hard to scan

![public-people-gatsby-news-date-publication-name](https://user-images.githubusercontent.com/235801/46971685-02d96c80-d0bd-11e8-9ae3-4d30d6d1f88e.png)

- memberships on mobile displace news so you usually have to scroll below the fold, and often further, to get to news, even when paging.

![public-people-gatsby-mobile-membership-displaces-news](https://user-images.githubusercontent.com/235801/46971686-03720300-d0bd-11e8-87c3-7cdd0941fa26.png)

- not clear whether membership dates are start or end dates. In django app this is not an issue because membership start/end are shown as points in time, like the articles, interleaved with them, while here they represent periods but we're showing single dates.

![public-people-gatsby-membership-dates](https://user-images.githubusercontent.com/235801/46971687-03720300-d0bd-11e8-8508-a38e204c8f72.png)
